### PR TITLE
Refactor snow worker spec generation code

### PR DESCRIPTION
**Pull request by jiayiwang7**
*Issue #, if available:*

https://github.com/aws/eks-anywhere/pull/4481#discussion_r1053468444

*Description of changes:*

Refactor snow worker spec builder with the `clusterapi.Workers` API.

*Testing (if applicable):*

Unit tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/providers/snow/workers.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+			KubeadmConfigTemplate:   kubeadmConfigTemplate,
+			ProviderMachineTemplate: machineTemplate,
+		})
+	}
+
+	if err := w.UpdateImmutableObjectNames(ctx, client, getMachineTemplate, MachineTemplateDeepDerivative); err != nil {
+		return nil, errors.Wrap(err, "updating snow worker immutable object names")
+	}
+
+	w.CAPASIPPools = capasPools
```
Comment (taneyland) [inline pkg/providers/snow/workers.go:73]: Just to confirm, so the IP pools are meant to just be another part of the workers spec, not tied to specific worker node groups?

Comment (jiayiwang7) [reply pkg/providers/snow/workers.go:73]: correct, not necessarily tied to worker node groups -- the same IPPool can be referenced by multiple worker groups.

File: pkg/providers/snow/workers.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+			Groups: make([]baseWorkerGroup, 0, len(spec.Cluster.Spec.WorkerNodeGroupConfigurations)),
+		},
+	}
+
+	capasPools := CAPASIPPools{}
+
+	for _, wc := range spec.Cluster.Spec.WorkerNodeGroupConfigurations {
+		machineConfig := spec.SnowMachineConfig(wc.MachineGroupRef.Name)
+
+		capasPools.addPools(machineConfig.Spec.Network.DirectNetworkInterfaces, spec.SnowIPPools)
```
Comment (g-gaston) [inline pkg/providers/snow/workers.go:51]: do we need to accumulate all the ip pools for all the worker node groups or should we reset it on each loop iteration?

Comment (jiayiwang7) [reply pkg/providers/snow/workers.go:51]: i chose accumulating because i can directly return capasPools in the end without another concat

**Pull request merged**